sqlair.go

```diff
@@ -92,15 +97,98 @@ type Iterator struct {
 	started bool
 }
 
+// A Least Recently Used Cache using a doubly linked list.
+type lruCache struct {
+	ll   *list.List
+	c    map[index]*list.Element
+	size int
+}
+
+// The cache key.
+type index struct {
+	tx *TX
+	s  *Statement
+}
+
+// The cache value.
+type entry struct {
+	key   index
+	value *sql.Stmt
+}
+
+func newLRUCache(size int) *lruCache {
+	c := &lruCache{}
+	c.ll = list.New()
+	c.c = make(map[index]*list.Element)
+	c.size = size
+	return c
+}
+
+func (c *lruCache) lookup(s *Statement, tx *TX) (*sql.Stmt, bool) {
+	e, ok := c.c[index{tx: tx, s: s}]
+	if !ok {
+		return nil, ok
+	}
+	c.ll.MoveToFront(e)
+	return e.Value.(*entry).value, ok
+}
+
+func (c *lruCache) add(s *Statement, tx *TX, ps *sql.Stmt) error {
+	k := index{tx: tx, s: s}
+	if e, ok := c.c[k]; ok {
+		c.ll.MoveToFront(e)
+		return nil
+	}
+	var err error
+	e := c.ll.PushFront(&entry{key: k, value: ps})
+	c.c[k] = e
+	if c.ll.Len() > c.size {
+		b := c.ll.Back()
+		delete(c.c, b.Value.(*entry).key)
+		c.ll.Remove(b)
+		// Close the prepared statement on removal from the cache.
+		err = b.Value.(*entry).value.Close()
+	}
+	return err
+}
+
+// Remove all statements prepared on tx from the cache.
+// Note that this does not close the sql.Stmt.
+func (c *lruCache) removeTX(tx *TX) {
+	for k, _ := range c.c {
+		if k.tx == tx {
+			delete(c.c, k)
+		}
+	}
+}
+
 // Query takes a context, prepared SQLair Statement and the structs mentioned in the query arguments.
 // It returns a Query object for iterating over the results.
 func (db *DB) Query(ctx context.Context, s *Statement, inputArgs ...any) *Query {
 	if ctx == nil {
 		ctx = context.Background()
 	}
 
-	qe, err := s.pe.Query(inputArgs...)
-	return &Query{qs: db.db, qe: qe, ctx: ctx, err: err}
+	// Query only actually prepares the query arguemnt, the sql already
+	// exists in pe. This and the prepared stmt lookup could be swapped.
+	var err error
+	var qe *expr.QueryExpr
+	qe, err = s.pe.Query(inputArgs...)
+	if err != nil {
+		return &Query{ctx: ctx, err: err}
+	}
+
+	db.prepedCacheMutex.Lock()
+	defer db.prepedCacheMutex.Unlock()
+	ps, ok := db.prepedCache.lookup(s, nil)
+	if !ok {
+		ps, err = db.db.PrepareContext(ctx, qe.QuerySQL())
+		if err == nil {
+			err = db.prepedCache.add(s, nil, ps)
+		}
+	}
+
+	return &Query{stmt: ps, qe: qe, ctx: ctx, err: err}
```

Comment:
This one return statement feels like it's tying together two paths, making it unclear at first glance which fields are and aren't `nil`. Perhaps it would be better to have a separate return in case the `db.prepedCache.add(...)` errors? This would also remove the need for a shared `var err error`.

Reply:
Changed to have a separate return for the `nil` case.